### PR TITLE
Try fixing ThreadResolveWorker calls

### PR DESCRIPTION
**Pull request by ClearlyClaire**
This is an attempt at fixing #3388.

From my understanding of ActiveRecord, a transaction is commited as soon as
the exit of the outmost ActiveRecord.transaction block. However, inner
transaction blocks will exit without the transaction being commited.

In this case, ThreadResolveWorker were fired *within* a transaction block,
so moving the call out of it should do the trick. However, this is somewhat
fragile, as this whole codepath could be called within yet another transaction.

### CODE REVIEW

**Comment by Gargron** [comment]
Oh, I see - the problem was that ThreadResolveWorker starts working before the transaction is committed, therefore it can't find the parent status it's supposed to act on. Good catch. Not sure if there's a downside to moving the in_reply_to_id setting out of the transaction. I mean, there is - bad result will get returned and cached from some of the public timeline API calls.

Ideally we would only move the async call out of the transaction, not the initial value if it's available...

**Comment by ThibG** [comment]
@Gargron hm, I don't think I'm changing the behaviour here. Only downside I see is that I'm making an extra `save!` call… maybe guarding it with `status.changed?` would make sense.

File: app/services/process_feed_service.rb [hunk 1]
```diff
@@ -72,6 +72,12 @@ def create_status
         status.save!
       end
 
+      if thread?(@xml)
+        Rails.logger.debug "Trying to attach #{status.id} (#{id(@xml)}) to #{thread(@xml).first}"
+        status.thread = find_or_resolve_status(status, *thread(@xml))
+        status.save!
```
Comment (ThibG) [inline app/services/process_feed_service.rb]: Maybe `status.save! if status.changed?` would make sense.

**Comment by Gargron** [comment]
@ThibG While we're inside the transaction block, calls to /api/v1/timelines/public will *not* return the status, because it doesn't exist yet. But it starts existing as soon as the transaction block commits. So committing it with an unset in_reply_to_id value will give clients an "early" version of the status, with an unset value. The caching layer will save it (though indeed, it should invalidate itself on next save! call)

**Comment by ThibG** [comment]
@Gargron ah I see. And even with the `changed?` that would make one avoidable `save!` call. The best thing to do is probably to only move the async call, yeah…

**Review by Gargron — APPROVED** [review]

File: app/services/process_feed_service.rb [hunk 2]
```diff
@@ -69,9 +69,16 @@ def create_status
           end
         end
 
+        status.thread = find_status(thread(@xml).first) if thread?(xml)
```
Comment (Gargron) [inline app/services/process_feed_service.rb]: Need `@xml` instead of `xml` here (Travis failure)

**Review by Gargron — CHANGES_REQUESTED** [review]

File: app/services/process_feed_service.rb (hunk 2 above)
Comment (ThibG) [inline app/services/process_feed_service.rb]: Oops, sorry you're right. Should be ok now!

**Review by Gargron — APPROVED** [review]

**Pull request merged**